csrc/type.h

```diff
@@ -469,6 +480,12 @@ inline bool isCompatibleDataType(DataType dtype, DataType dtype2) {
     }
     return true;
   }
+  if (std::holds_alternative<OpaqueType>(dtype.type) &&
+      std::holds_alternative<OpaqueType>(dtype2.type)) {
+    const auto& opaque_type = std::get<OpaqueType>(dtype.type);
+    const auto& opaque_type2 = std::get<OpaqueType>(dtype2.type);
+    return opaque_type.type_info.get() == opaque_type2.type_info.get();
```

Comment:
I'm not entirely sure what "compatible" means in this context. If the underlying type of `dtype2` is a subclass of that of `dtype`, will this return `true`? Should it?

Reply:
This is a good question, and we don't have a clear answer yet. In the past, we were treating all opaques equivalent, that is `Opaque{123}` and `Opaque{Fusion{}}` are the same dtype, which definitely doesn't make sense. With this change, we are being the strictest: we require the dtypes to be exactly the same for two opaques to be considered equal type. I don't think we have any use case that needs us to handle subclass structures, but in the future, if we do, I think we should make it a separate dtype, instead of making it an Opaque.